### PR TITLE
Expose swarm object on tracker server

### PR DESCRIPTION
**Pull request by yciabaud**
This PR allows a server tracker to intercept announce request and responses. I use it to provide extra data to peers.

I don't expect this PR to be reviewed either, I post this in case someone is interested.
Tell me if you don't want PRs like that anymore...

I tried to clarify this in #207 but I did not have any clue after 2 weeks.

### CODE REVIEW

File: test/request-handler.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+var Buffer = require('safe-buffer').Buffer
```
Comment (feross) [inline test/request-handler.js]: Thanks for adding tests.

**Comment by yciabaud** [comment]
Thank you @feross for the review, you right this might be bad designed and this is working only with websockets.

In my context, I have some peers that I want to seed their files via HTTP to avoid some WebRTC issues (WebRTC not supported, resources ...). I need the other peers to add them as webseeds and not as webrtc peers.

To achieve this, I add an http port number to the announce fields in the client and my tracker must send the data back to the swarm.

I am using this interceptor to push the webseed urls in the swarm on the ws channels.

How do you think this could be handled? Those are peers too so I believe I should track them in the torrent swarm but I cannot make it fit the bittorrent protocol (just like ws peers have different fields).

Maybe we could support webseeds peers besides WebRTC peers because it can make sense in the browser, but I felt like it should be implemented as an extension in the websocket server.

What do you think ? How should I handle this? Should bittorrent-tracker allow users to extends the websocket protocol ? Is this webseed tracking thing worth to be added to webtorrent ?

**Comment by feross** [comment]
Thanks for explaining. I'm not sure I understand 100% why this is necessary, so let me ask some additional questions:

1. Are you building an app where some of your peers can actually run HTTP servers and act as web seeds in the p2p network? I assume those peers are running in Node.js. Is this correct?

1. Is it possible to figure out which peers will be web seeds in advance? Because if it is possible, then you could simply embed the webseed URLs in the magnet link or torrent file. Then, no changes to the tracker server are necessary. This is the easiest way to use webseeds.

1. If you require the ability for a peer to dynamically add themselves to this "webseed list", then is it necessary to implement this as an extension to the tracker protocol? If you only intend to use this feature for your own app, then it could be implemented any way you like, outside of the tracker server. For example, you could add an HTTP endpoint `GET /webseeds` that would return a list of webseed URLs. And peers could `POST /webseeds` to add themselves to the "webseed list".

It's still possible to change the tracker to support these options, but I want to make sure you've considered these other options, which seem easier and cleaner to me.

**Comment by yciabaud** [comment]
> Are you building an app where some of your peers can actually run HTTP servers and act as web seeds in the p2p network? I assume those peers are running in Node.js. Is this correct?

You're right, I do have a Node.js client running a webserver and acting as seeds. 
 
> Is it possible to figure out which peers will be web seeds in advance? Because if it is possible, then you could simply embed the webseed URLs in the magnet link or torrent file. Then, no changes to the tracker server are necessary. This is the easiest way to use webseeds.

Those peers are not running on servers, which means some files can be deleted or the peers can suddenly go offline. I would like to add/remove new peers dynamically.

> If you require the ability for a peer to dynamically add themselves to this "webseed list", then this more complicated. Consider: Is it necessary to implement this as an extension to the tracker protocol? If you only intend to use this feature for your own app, then it could be implemented any way you like outside of the tracker server. For example, you could add an HTTP endpoint GET /webseeds that would return a list of webseed URLs. And peers could POST /webseeds to add themselves to the "webseed list".

I don't know if this is has to be an extension to the protocol, that is why my PR only tries to let an app using bittorent-tracker like mine to extends the websocket protocol for its personal needs.

My point is that I have an app communicating with clients for tracking purpose using websockets and I would like to use this connection to provide extra tracking features.
Do you think we can propose another way to allow server users to communicate with their peers?
Otherwise I guess I can find the websocket in the swarm and push the messages on my own...

**Comment by feross** [comment]
I think that we already have an official way to support "intercepting" announces, though it's not documented. We added it back when BitTorrent, Inc. wanted to use this library as the tracker for BitTorrent Bundles.

Basically, `Server.prototype.createSwarm` is designed to be overwritten, so you can specify your own Swarm implementation. With ES6 classes, it's even easier to overwrite just the `announce()` method.

If we expose the `Swarm` implementation on `Server.Swarm`, you could do:

```js
class Swarm extends Server.Swarm {
  announce (params, cb) {
    super.announce(params, function (err, response) {
      if (err) return cb(response)
      response.complete = 246
      response.extraData = 'hi'
      cb(null, response)
    })
  }
}

Server.Swarm = Swarm

var myServer = new Server(/* ... */)
```

It's a bit lower-level, but I think this works. I'll send a PR to update this PR with these changes.

**Comment by feross** [comment]
PR to this PR is here: https://github.com/yciabaud/bittorrent-tracker/pull/1

**Comment by yciabaud** [comment]
This is a cleaner way to intercept requests, we need to overwrite things but I believe it is acceptable in the context of extending the protocol for personal needs.

**Comment by feross** [comment]
Released as 9.1.0.

**Pull request merged**